File: internal/git/git.go
// Package git provides helper functions for common git operations.
package git

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"strings"
)

// Clone runs git clone <url> <dir>. If sshKeyPath is non-empty, it's used
// as the identity file for SSH authentication.
func Clone(ctx context.Context, url, dir, branch, sshKeyPath string) error {
	args := []string{"clone", "--branch", branch, "--single-branch", url, dir}
	cmd := exec.CommandContext(ctx, "git", args...)
	if sshKeyPath != "" {
		cmd.Env = append(os.Environ(),
			fmt.Sprintf("GIT_SSH_COMMAND=ssh -i %s -o StrictHostKeyChecking=accept-new", sshKeyPath))
	}
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("git clone: %w: %s", err, string(out))
	}
	return nil
}

// IsRepo checks if the directory is a git repository.
func IsRepo(dir string) bool {
	cmd := exec.Command("git", "-C", dir, "rev-parse", "--git-dir")
	return cmd.Run() == nil
}

// ConfigureSSHKey sets the core.sshCommand in the repo's local git config
// so all git operations use the specified SSH key.
func ConfigureSSHKey(repoDir, sshKeyPath string) error {
	cmd := exec.Command("git", "config", "core.sshCommand",
		fmt.Sprintf("ssh -i %s -o StrictHostKeyChecking=accept-new", sshKeyPath))
	cmd.Dir = repoDir
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("git config core.sshCommand: %w: %s", err, string(out))
	}
	return nil
}

// HasRemote checks if the repo has an "origin" remote configured.
func HasRemote(ctx context.Context, repoDir string) bool {
	cmd := exec.CommandContext(ctx, "git", "remote", "get-url", "origin")
	cmd.Dir = repoDir
	return cmd.Run() == nil
}

// Fetch runs git fetch origin <branch> in the given directory.
func Fetch(ctx context.Context, repoDir, branch string) error {
	cmd := exec.CommandContext(ctx, "git", "fetch", "origin", branch)
	cmd.Dir = repoDir
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("git fetch origin %s: %w: %s", branch, err, string(out))
	}
	return nil
}

// Checkout runs git checkout <ref> in the given directory.
func Checkout(ctx context.Context, repoDir, ref string) error {
	cmd := exec.CommandContext(ctx, "git", "checkout", ref)
	cmd.Dir = repoDir
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("git checkout %s: %w: %s", ref, err, string(out))
	}
	return nil
}

// CurrentSHA runs git rev-parse HEAD and returns the current SHA.
func CurrentSHA(ctx context.Context, repoDir string) (string, error) {
	cmd := exec.CommandContext(ctx, "git", "rev-parse", "HEAD")
	cmd.Dir = repoDir
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("git rev-parse HEAD: %w: %s", err, string(out))
	}
	return strings.TrimSpace(string(out)), nil
}

// LsRemote runs git ls-remote to verify access to a repository.
func LsRemote(ctx context.Context, url, sshKeyPath string) error {
	cmd := exec.CommandContext(ctx, "git", "ls-remote", "--heads", url)
	if sshKeyPath != "" {
		cmd.Env = append(os.Environ(),
			fmt.Sprintf("GIT_SSH_COMMAND=ssh -i %s -o StrictHostKeyChecking=accept-new", sshKeyPath))
	}
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("git ls-remote: %w: %s", err, string(out))
	}
	return nil
}

// SetRemoteURL updates the origin remote URL.
func SetRemoteURL(ctx context.Context, repoDir, url string) error {
	cmd := exec.CommandContext(ctx, "git", "remote", "set-url", "origin", url)
	cmd.Dir = repoDir
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("git remote set-url: %w: %s", err, string(out))
	}
	return nil
}

// RemoteSHA runs git rev-parse origin/<branch> and returns the remote branch SHA.
func RemoteSHA(ctx context.Context, repoDir, branch string) (string, error) {
	ref := "origin/" + branch
	cmd := exec.CommandContext(ctx, "git", "rev-parse", ref)
	cmd.Dir = repoDir
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("git rev-parse %s: %w: %s", ref, err, string(out))
	}
	return strings.TrimSpace(string(out)), nil
}

File: src/modules/sources/git.ts
import { $ } from '@/libs/shell'
import { InternalError } from '@jib/errors'

interface ShellOutput {
  exitCode: number | null
  stdout: string
  stderr: string
}

export interface GitEnv {
  GIT_SSH_COMMAND?: string
}

/** Distinguishes zx process output from a Jib error; zx output also extends Error. */
function isShellOutput(value: ShellOutput | InternalError): value is ShellOutput {
  return typeof value === 'object' && value !== null && 'exitCode' in value
}

/** Maps a non-zero Git process result to an internal error. */
function commandFailure(
  out: ShellOutput | InternalError,
  label: string,
): InternalError | undefined {
  if (!isShellOutput(out)) {
    return out
  }
  if (out.exitCode === 0) {
    return undefined
  }
  const detail =
    out.stderr.toString().trim() ||
    out.stdout.toString().trim() ||
    `command exited with code ${out.exitCode}`
  return new InternalError(detail ? `${label}: ${detail}` : label)
}

/** Runs one Git command without throwing for a non-zero process exit. */
async function run(
  args: string[],
  env: GitEnv = {},
  dir?: string,
): Promise<ShellOutput | InternalError> {
  const mergedEnv = { ...process.env, ...env } as Record<string, string>
  const command = dir ? ['git', '-C', dir, ...args] : ['git', ...args]
  try {
    return await $({ env: mergedEnv, quiet: true })`${command}`
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error)
    return new InternalError(`git ${args[0] ?? 'command'}: ${message}`, { cause: error })
  }
}

/** Clones a repo into `dir`, returning a typed git error on failure. */
export async function sourcesGitClone(
  url: string,
  dir: string,
  opts: { branch?: string; env?: GitEnv } = {},
): Promise<InternalError | undefined> {
  const args = ['clone']
  if (opts.branch) {
    args.push('--branch', opts.branch, '--single-branch')
  }
  args.push(url, dir)
  const failure = commandFailure(await run(args, opts.env ?? {}), 'git clone')
  if (failure) {
    return failure
  }
  await markSafeDirectory(dir)
}

/** Adds a shared checkout to Git's safe-directory list as a best-effort compatibility step. */
async function markSafeDirectory(dir: string): Promise<void> {
  await run(['config', '--global', '--add', 'safe.directory', dir])
}

/** Fetches `ref` from `origin`, or all refs when `ref` is empty. */
export async function sourcesGitFetch(
  dir: string,
  ref = '',
  env: GitEnv = {},
): Promise<InternalError | undefined> {
  const args = ref ? ['fetch', 'origin', ref] : ['fetch', 'origin']
  return commandFailure(await run(args, env, dir), 'git fetch')
}

/** Checks out `ref` in an existing local checkout. */
export async function sourcesGitCheckout(
  dir: string,
  ref: string,
): Promise<InternalError | undefined> {
  return commandFailure(await run(['checkout', ref], {}, dir), `git checkout ${ref}`)
}

/** Returns the current `HEAD` SHA for a local checkout. */
export async function sourcesGitCurrentSha(dir: string): Promise<string | InternalError> {
  const result = await run(['rev-parse', 'HEAD'], {}, dir)
  const failure = commandFailure(result, 'git rev-parse')
  if (failure) {
    return failure
  }
  if (!isShellOutput(result)) {
    return result
  }
  return result.stdout.toString().trim()
}

/** Returns the SHA recorded in `FETCH_HEAD` after a fetch. */
export async function sourcesGitFetchedSha(dir: string): Promise<string | InternalError> {
  const result = await run(['rev-parse', 'FETCH_HEAD'], {}, dir)
  const failure = commandFailure(result, 'git rev-parse FETCH_HEAD')
  if (failure) {
    return failure
  }
  if (!isShellOutput(result)) {
    return result
  }
  return result.stdout.toString().trim()
}

/** Resolves the SHA for `ref` without creating a local checkout. */
export async function sourcesGitLsRemote(
  url: string,
  ref = 'HEAD',
  env: GitEnv = {},
): Promise<string | InternalError> {
  const result = await run(['ls-remote', url, ref], env)
  const failure = commandFailure(result, 'git ls-remote')
  if (failure) {
    return failure
  }
  if (!isShellOutput(result)) {
    return result
  }
  const first = result.stdout.toString().trim().split('\n')[0] ?? ''
  return (first.split('\t')[0] ?? '').trim()
}

/** Resolves the remote default branch, if the remote reports one. */
export async function sourcesGitDefaultBranch(
  url: string,
  env: GitEnv = {},
): Promise<string | InternalError | undefined> {
  const result = await run(['ls-remote', '--symref', url, 'HEAD'], env)
  const failure = commandFailure(result, 'git ls-remote --symref')
  if (failure) {
    return failure
  }
  if (!isShellOutput(result)) {
    return result
  }
  for (const line of result.stdout.toString().split('\n')) {
    const match = /^ref:\s+refs\/heads\/([^\t]+)\tHEAD$/.exec(line.trim())
    if (match?.[1]) {
      return match[1]
    }
  }
  return undefined
}

/** Returns true when `dir` is already a Git worktree. */
export async function sourcesGitIsRepo(dir: string): Promise<boolean> {
  const result = await run(['rev-parse', '--git-dir'], {}, dir)
  return isShellOutput(result) && result.exitCode === 0
}

/** Returns true when `origin` is configured for `dir`. */
export async function sourcesGitHasRemote(dir: string): Promise<boolean> {
  const result = await run(['remote', 'get-url', 'origin'], {}, dir)
  return isShellOutput(result) && result.exitCode === 0
}

/** Updates the configured `origin` URL for an existing checkout. */
export async function sourcesGitSetRemoteUrl(
  dir: string,
  url: string,
): Promise<InternalError | undefined> {
  return commandFailure(
    await run(['remote', 'set-url', 'origin', url], {}, dir),
    'git remote set-url',
  )
}

/** Builds an SSH env override that pins Git to the chosen private key. */
export function sourcesGitConfigureSshKey(keyPath: string): GitEnv {
  return {
    GIT_SSH_COMMAND: `ssh -i ${keyPath} -o StrictHostKeyChecking=accept-new`,
  }
}
